polaris/evaluate/_metric.py

```diff
@@ -84,55 +164,241 @@ class Metric(Enum):
+    """
+
+    kind: Literal["grouped"] = "grouped"
+    config: GroupedMetricConfig
+
+    @model_validator(mode="after")
+    def check_default_required(self) -> Self:
+        """Check that a default value is provided when on_error is 'default'."""
+        handling = self.config.on_error
+        default = self.config.default
+        if handling == "default" and default is None:
+            raise ValueError("Default value must be provided when on_error is 'default'")
+        return self
+
+    def score(
+        self,
+        y_true: GroundTruth,
+        y_pred: BenchmarkPredictions | None = None,
+        y_prob: BenchmarkPredictions | None = None,
     ) -> float:
-        """For convenience, make metrics callable"""
-        return self.score(y_true, y_pred, y_prob)
+        """Compute the metric.
+
+        Args:
+            y_true: The true target values.
+            y_pred: The predicted target values, if any.
+            y_prob: The predicted target probabilities, if any.
+        """
+
+        pred = prepare_predictions(y_pred, y_prob, self.y_type)
+
+        # NOTE (cwognum): We rely on pandas to due the grouping.
+        # This does imply a memory bottleneck.
+        df = y_true.as_dataframe()
+        df[self.y_type] = pred.flatten()
+
+        # Compute the metric for each group
+        scores = []
+
+        for _, group in df.groupby(self.config.group_by):
+            y_true_group = group[y_true.target_cols[0]].values
+            y_pred_group = group[self.y_type].values
+
+            kwargs = prepare_metric_kwargs(
+                y_true_group,
+                y_pred_group,
+                self.y_type,
+            )
+
+            # A group can be arbitrarily small (e.g. only 1 item), which also has implications
+            # for its target distribution (e.g. only the positive class). These conditions can
+            # lead to errors in the metric computation, which we handle here.
+            try:
+                score = self.fn(**kwargs, **self.metric.value.kwargs)
+            except Exception:
+                if self.config.on_error == "ignore":
+                    continue
+                elif self.config.on_error == "raise":
+                    raise
+                score = self.config.default
+            scores.append(score)
+
+        # Aggregate the per-group score in a single score.
+        if self.config.aggregation == "mean":
+            score = np.mean(scores)
+        else:
+            score = np.median(scores)
+        return score
+
+
+# Keeping this code in this file for now because it's closely coupled with the Metric classes above.
+
+MetricType: TypeAlias = Metric | GroupedMetric
+"""
+A union of all supported Metric types in Polaris.
+"""
+
+AnnotatedMetricType = Annotated[MetricType, Field(..., discriminator="kind")]
+"""
+Annotates the MetricType with a discriminator field.
+"""
+
+
+def instantiate_metric(m: MetricType | str | dict):
```

Comment:
**Comment**
You don't need this anymore, or the union types above, if it's all in one `Metric` model.